api/envoy/extensions/rbac/matchers/upstream_ip/v3/upstream_ip_matcher.proto

```diff
@@ -0,0 +1,29 @@
+syntax = "proto3";
+
+package envoy.extensions.rbac.matchers.upstream_ip.v3;
+
+import "envoy/config/core/v3/address.proto";
+
+import "udpa/annotations/status.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.rbac.matchers.upstream_ip.v3";
+option java_outer_classname = "UpstreamIpMatcherProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: RBAC upstream IP matcher plugin]
+// [#extension: envoy.rbac.matchers.upstream_ip]
+
+// This is configuration for matching upstream ip.
+message UpstreamIpMatcher {
+  // A CIDR block that will be used to match the upstream IP.
+  // This matcher requires a filter in the chain to have saved the upstream address(es) in the
+  // filter state before the matcher is executed by RBAC filter. The state should be saved with key
+  // `envoy.stream.upstream_address_set` (See
```

Comment:
OK, looks much better now. Two other nits:
1. I don't think the data structure matters, I would just call this `envoy.stream.upstream_addresses`.
2. What are the semantics for match? Presumably on DENY rules "none of the addresses match this CIDR range", what about with ALLOW rules?

Reply:
@htuch Thanks for review again.

 1. I will update the API for 2
 2. Do you want the semantics documented? 
 Wanted to clarify here: 
 For ALLOW rules, if the traffic is ALLOWED IFF the match succeeds. For example, if the rules are ALLOW upstream_ip 10.10.1.0/24 OR 192.168.0.0/16, then if the upsream IP matches EITHER of them , only then the upstream connection be made. For AND rules, this could be used by combining IP and PORT for example. So we can have policies which allows traffic IFF upstream IP AND upstream port matches.

Do you want the above documented ?

Reply:
I'm not sure I fully grok the match semantics from (2) above. You have a bag of addresses. Are the semantics that the matcher matches IFF there exists an address in `envoy.stream.upstream_address` that matches an address in the CIDR range of `upstream_ip`? This could be a cleaner way to put it.

These are weird semantics for ALLOW though; for DENY they make total sense, since you want to reject any IP that might cause problems. However, when ALLOWing by IP, you are typically allowlisting, so you want to ensure that all resolved IPs match the allowlist.

Reply:
> I'm not sure I fully grok the match semantics from (2) above. You have a bag of addresses. Are the semantics that the matcher matches IFF there exists an address in `envoy.stream.upstream_address` that matches an address in the CIDR range of `upstream_ip`? This could be a cleaner way to put it.

This is correct.


> These are weird semantics for ALLOW though; for DENY they make total sense, since you want to reject any IP that might cause problems. However, when ALLOWing by IP, you are typically allowlisting, so you want to ensure that all resolved IPs match the allowlist.

Yes the matcher currently has ANY semantics. That is - it will match  `true` if ANY of the ip address in the bag is a match. We can improve on this by adding another attribute like `match_type = ANY` (along with `match_type = OR`) to enhance the current semantics.

Reply:
This now makes sense to me, but port match is a bit confusing :/